protocol_details/rpmsg_protocol.rst

```diff
@@ -0,0 +1,89 @@
+
+
+The use of interrupts by the ring buffers is optional and can be suppressed using a bit set in their configuration flags field - in such a configuration, the interrupts are not necessary. However both cores need to poll these ring buffers for new incoming messages, which may not be optimal.
+
+.. image:: ../images/core_to_core_interrupt.jpg
+
+Media Access Layer - VirtIO
+---------------------------
+
+This layer is the key part of the whole solution - thanks to this layer, there is no need for inter-core synchronization. This is achieved by a technique called single-writer single-reader circular buffering, which is a data structure enabling multiple asynchronous contexts to interchange data.
+
+.. image:: ../images/vrings.jpg
+
+This technique is however applicable only in core-to-core configuration, not in core-to-multicore configuration, since in such a case, there would be multiple writers to the “IN” ring buffer. This would require a synchronization element, [such as a semaphore?], which is not desirable.
+
+The above shown picture describes the vring component. Vring is composed of three elementary parts - buffer descriptor pool, the “available” ring buffer (or input ring buffer) and the “used” ring buffer (or free ring buffer). All three elements are physically stored in the shared memory.
+
+Each buffer descriptor contains a 64-bit buffer address, which holds an address to a buffer stored in the shared memory (as seen physically by the “receiver” or host of this vring), its length as a 32-bit variable, 16-bit flags field and 16-bit link to the next buffer descriptor. The link is used to chain unused buffer descriptors and to chain descriptors, which have the F_NEXT bit set in the flags field to the next descriptor in the chain.
+
+Note that the available and used ring buffer areas contain pointers to buffers, rather than the buffers themselves. Later versions of the `Virtio virtqueue <https://docs.oasis-open.org/virtio/virtio/v1.3/csd01/virtio-v1.3-csd01.html#x1-270006>`_ have change naming of these three sections to Descriptor Table, Driver Area and Device Area.
+
+.. image:: ../images/vring_descriptor.jpg
+
+The input ring buffer contains its own flags field, where only the 0th bit is used - if it is set, the “writer” side should not be notified, when the “reader” side consumes a buffer from the input or “avail” ring buffer. By default the bit is not set, so after the reader consumes a buffer, the writer should be notified by triggering an interrupt. The next field of the input ring buffer is the index of the head, which is updated by the writer, after a buffer index containing a new message is written in the ring[x] field.
+
+.. image:: ../images/vring_descriptor_flags.jpg
```

Comment:
It's not clear from the title of this image that it refers to the available ring buffer.